### PR TITLE
CLDR-14755 add tools for checking grammar results

### PR DESCRIPTION
**Pull request by macchiati**
CLDR-14755

This also makes some improvements to the ExampleGenerator that will show up in the Survey Tool. Notably, the selection of samples for the Minimal pairs is improved, to pick cases that show the best variation (eg units that have the largest number of distinctive forms).

BestMinimalPairSamples.java 
is now a centralized source for getting samples for insertion into minimal pairs

ExampleGenerator.java 
The logic for the examples is now simplified to use BestMinimalPairSamples

ListGrammarData.java
A tool for gathering information across locales for inspection

ChartGrammaticalForms.java
This has a bit of work, but will have more in subsequent PRs.

- [ ] This PR completes the ticket.

### CODE REVIEW

File: tools/cldr-code/src/main/java/org/unicode/cldr/test/BestMinimalPairSamples.java [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+                    genderResults.put(possibleGender, longUnitId);
+                }
+            }
+            if (!unitCases.isEmpty()) {
+                unitPatternToCases.clear();
+                for (String count : pluralInfo.getKeywords()) {
+                    for (String unitCase : unitCases) {
+                        String grammarAttributes = GrammarInfo.getGrammaticalInfoAttributes(grammarInfo, UnitPathType.unit, count, null, unitCase);
+                        String unitPattern = cldrFile.getStringValue("//ldml/units/unitLength[@type=\"long\"]/unit[@type=\"" + longUnitId + "\"]/unitPattern" + grammarAttributes);
+                        if (unitPattern == null) {
```
Comment (btangmu) [inline tools/cldr-code/src/main/java/org/unicode/cldr/test/BestMinimalPairSamples.java:135]: deep for/if nesting could also be improved by more classes/methods

File: tools/cldr-code/src/main/java/org/unicode/cldr/test/ExampleGenerator.java [hunk 2]
```diff
@@ -647,86 +646,40 @@ private String handleDayPeriod(XPathParts parts, String value) {
...
         case "genderMinimalPairs": //ldml/numbers/minimalPairs/genderMinimalPairs[@gender="feminine"]
             String gender = parts.getAttributeValue(-1, "gender");
-            //ldml/units/unitLength[@type="long"]/unit[@type="duration-day"]/unitPattern[@count="one"]
-            shortUnitId = getBestUnitWithGender(gender);
-            //unitPattern = cldrFile.getStringValue("//ldml/units/unitLength[@type=\"long\"]/unit[@type=\"" + shortUnitId + "\"]/displayName");
-            // we do the following to get a singular form, since we can't depend on the displayName to be singular
-            unitPattern = cldrFile.getStringValue("//ldml/units/unitLength[@type=\"long\"]/unit[@type=\"" + shortUnitId + "\"]/unitPattern[@count=\"" + count + "\"]");
-            unitPattern = unitPattern.replace("{0}", "").trim();
-            //ldml/units/unitLength[@type="long"]/unit[@type="duration-day"]/gender
+            sample = bestMinimalPairSamples.getBestUnitWithGender(gender, output);
```
Comment (btangmu) [inline tools/cldr-code/src/main/java/org/unicode/cldr/test/ExampleGenerator.java:673]: this encapsulation is a great improvement

**Review by btangmu — DISMISSED** [review]
LGTM, without spending the time to study in detail; made a few suggestions for enhancing maintainability

**Comment by btangmu** [comment]
@macchiati per your email I'll try to resolve the conflicts

**Review by btangmu — APPROVED** [review]

**Review by btangmu — APPROVED** [review]

**Comment by btangmu** [comment]
I after resolving the conflicts, my commit gets "Commit message for a2754fb fails validation". I clicked on Details and tried to edit the message and squash, but I got a 404 error and a message saying I might not have permission.

There's also a message "As an administrator, you may still merge this pull request" and I plan to do that if the tests pass...

**Comment by macchiati** [comment]
Thanks!

**Comment by btangmu** [comment]
Error summary:
CLDR/TestAnnotations/TestUniqueness
CLDR/TestAttributeValues/TestValid
CLDR/TestDayPeriods/TestAttributes
CLDR/TestExampleGenerator/TestMinimalPairExamples
CLDR/TestUnits/TestGender

One of the errors:
 TestAnnotations {
...
    TestUniqueness
Error:  (TestAnnotations.java:233) Duplicate name in hy: “անձ․ միջին մուգ գույնի մաշկ և շիկահեր” for 👱🏾 & 🧑🏾‍🦰

**Comment by btangmu** [comment]
@macchiati I'm guessing we want to merge in spite of the errors, and to fix them in xml -- could you please confirm? Or do the squash and merge yourself, whichever you prefer?

Also I imagine "Commit message for a2754fb fails validation" isn't a show-stopper though I wish I knew how to fix it...

**Pull request merged**